include/revng/Pipeline/Container.h

```diff
@@ -0,0 +1,74 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <memory>
+
+#include "revng/Pipeline/Target.h"
+
+namespace Pipeline {
+
+class ContainerBase {
+private:
+  const char *ID;
+
+public:
+  const char *getTypeID() const { return ID; }
+
+  ContainerBase(char const *ID) : ID(ID) {}
+
+  /// The implementation of cloneFiltered must return a copy of the current
+  /// container and a invocation of enumerate on such container must be exactly
+  /// equal to Targets
+  virtual std::unique_ptr<ContainerBase>
+  cloneFiltered(const TargetsList &Targets) const = 0;
+
+  /// The implementation of this method must ensure that after the execution
+  /// this->enumerate() == before(Other).enumerate().merge(this->enumerate())
+  ///
+  /// In other words the whole content of Other must be transferred to this.
+  virtual void mergeBack(ContainerBase &&Other) = 0;
+
+  /// This method is the method used the pipeline system to understand which
+  /// targets are currently aviable inside the current container.
+  virtual TargetsList enumerate() const = 0;
+
+  /// The implementation must ensure that
+  /// not after(this)->enumerate().contains(Targets);
+  virtual bool remove(const TargetsList &Targets) = 0;
+
+  /// The implementation must ensure that there exists a file at the provided
+  /// path that contains the serialized version of this object.
+  virtual llvm::Error storeToDisk(llvm::StringRef Path) const = 0;
+
+  /// The  implementation must esure that the content of this file will be
+  /// loaded from the provided path.
+  virtual llvm::Error loadFromDisk(llvm::StringRef Path) = 0;
+  virtual ~ContainerBase() = default;
+
+  static bool classof(const ContainerBase *) { return true; }
+};
+
+/// crtp class to be extended to implement a pipeline container
+/// The methods that must be implemented are those shown in containerBase
+template<typename Derived>
+class Container : public ContainerBase {
+public:
+  Container() : ContainerBase(&Derived::ID) {}
+
+  void mergeBack(ContainerBase &&Container) final {
+    mergeBackImpl(std::move(llvm::cast<Derived>(Container)));
+  }
+  ~Container() override = default;
+
+  static bool classof(const ContainerBase *Base) {
+    return Base->getTypeID() == &Derived::ID;
+  }
+
+protected:
+  virtual void mergeBackImpl(Derived &&Container) = 0;
+};
```

Comment:
At this point the role of this class is really minimal.
In an effort to reduce layers of abstractions we might consider removing it and push its features into inheritors.

Reply:
i would prefer keeping it, because this is used in tests and containers like TemporaryFile Container

Reply:
An extra abstraction layer to understand.
If you'd like to keep then:
1. add a concept on `Derived` (it needs to have `::ID`)
2. move the `self` methods in here
3. make sure `mergeBackImpl` is not public in inheritors